### PR TITLE
Feat/dataroom linktree

### PR DESCRIPTION
**Pull request by digant2482**
#69. @mfts. I have disabled the ability to create hierarchical datarooms. Please let me know if any changes are needed.

### CODE REVIEW

**Comment by Aashish-Upadhyay-101** [comment]
I will review this PR @digant2482 
cc @mfts

**Review by Aashish-Upadhyay-101 — COMMENTED** [review]
`datarooms` would be good instead of `dataroom`

File: prisma/schema.prisma [hunk 1]
```diff
@@ -47,6 +47,7 @@ model User {
   accounts       Account[]
   sessions       Session[]
   documents      Document[]
+  dataroom       Dataroom[]
```
Comment (Aashish-Upadhyay-101) [inline prisma/schema.prisma]: `datarooms` would be good instead of `dataroom`

**Comment by Aashish-Upadhyay-101** [comment]
hey @digant2482 I can still see some hierarchical's data room's code! 
could you clean those up first before I dive deeper into your code?

**Comment by Aashish-Upadhyay-101** [comment]
hi @digant2482 please respond!

File: prisma/schema.prisma (hunk 1 above)
Comment (digant2482) [reply prisma/schema.prisma]: On it

**Comment by digant2482** [comment]
> hey @digant2482 I can still see some hierarchical's data room's code! could you clean those up first before I dive deeper into your code?

Hi @Aashish-Upadhyay-101, Noted

**Comment by digant2482** [comment]
> nice, @digant2482 ping me, after you are done cleaning up !

Hey Ashish, all set, please let me know, you need any help

**Comment by Aashish-Upadhyay-101** [comment]
awesome !

File: pages/datarooms/index.tsx [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+        if (!response.ok) {
+          return;
+        }
+
+        const data = await response.json();
+        setDatarooms(data.datarooms);
+      }
+    })();
+  }, [teamInfo]);
```
Comment (Aashish-Upadhyay-101) [inline pages/datarooms/index.tsx]: please use useSWR hook to fetch data! See other examples

File: pages/view/[linkId]/index.tsx [hunk 3]
```diff
@@ -107,6 +109,7 @@ export default function ViewPage({
     email: string;
   };
   const { data: session, status } = useSession();
+  const { authenticationCode } = router.query as { authenticationCode: string };
```
Comment (Aashish-Upadhyay-101) [inline pages/view/[linkId]/index.tsx]: `authenticationCode` isn't used in this component! You can remove this.

Comment (digant2482) [reply pages/view/[linkId]/index.tsx]: Done

File: pages/datarooms/index.tsx (hunk 2 above)
Comment (digant2482) [reply pages/datarooms/index.tsx]: Done

**Comment by digant2482** [comment]
Hi @Aashish-Upadhyay-101, suggested changes are done

**Comment by Aashish-Upadhyay-101** [comment]
@digant2482  hey, did you scoped the Dataroom by Teams?

**Comment by digant2482** [comment]
> @digant2482 hey, did you scope the Dataroom by Teams?

Yes

**Comment by Aashish-Upadhyay-101** [comment]
I noticed that the `dataroom` endpoints are outside the team folder in the api,
and you are passing `teamId` via `req.body`
could you please put the dataroom inside team folder and get the team's id from `req.query` ?

**Comment by digant2482** [comment]
> I noticed that the `dataroom` endpoints are outside the team folder in the api, and you are passing `teamId` via `req.body` could you please put the dataroom inside team folder and get the team's id from `req.query` ?

Hi, Ashish, dataroom was not put inside teams folder because the end users also gets their datarooms from the same API (to avoid code duplication), now the end users won't have teamId, hence it is outside. Please let me know if you need more information

File: lib/swr/use-datarooms.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  const teamInfo = useTeam();
+  const { data: datarooms, error } = useSWR<
+    DataroomWithFilesFoldersAndFilesCount[]
+  >(
+    teamInfo?.currentTeam?.id &&
+      `/api/datarooms?teamId=${teamInfo?.currentTeam?.id}`,
+    fetcher,
+    {
+      revalidateOnFocus: false,
+      dedupingInterval: 10000,
```
Comment (Aashish-Upadhyay-101) [inline lib/swr/use-datarooms.ts]: I think this should be a bit more, like 20seconds+

File: lib/swr/use-dataroom.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    dataroomId: string;
+    authenticationCode: string | undefined;
+  };
+
+  // only fetch data once when dataroomId is present
+  const { data: dataroom, error } = useSWR<DataroomWithFiles>(
+    dataroomId && `/api/datarooms/paged?id=${encodeURIComponent(dataroomId)}`,
+    fetcher,
+    {
+      dedupingInterval: 10000,
```
Comment (Aashish-Upadhyay-101) [inline lib/swr/use-dataroom.ts]: same here 20seconds+

File: components/view/datarooms/dataroom-view.tsx [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+        onSubmitHandler={handleSubmit}
+        requireEmail={emailProtected}
+        requirePassword={!!dataroomPassword}
+        isLoading={isLoading}
+      />
+    );
+  }
+
+  if (isLoading) {
+    console.log("loading");
```
Comment (Aashish-Upadhyay-101) [inline components/view/datarooms/dataroom-view.tsx:211]: please remove unwanted console.logs

File: components/view/datarooms/dataroom-view.tsx [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+        onSubmitHandler={handleSubmit}
+        data={data}
+        isLoading={isLoading}
+      />
+    );
+  }
+
+  // If link is not submitted and does not have email / password protection, show the access form
+  if (!submitted && isProtected) {
+    console.log("calling access form");
```
Comment (Aashish-Upadhyay-101) [inline components/view/datarooms/dataroom-view.tsx]: remove this too

**Comment by Aashish-Upadhyay-101** [comment]
I have almost reviewed your code, I just have one thing to ask,
you added a new verification system, could you explain me about why you add it and how you implemented it? `email-authcode`

**Comment by digant2482** [comment]
> I have almost reviewed your code, I just have one thing to ask, you added a new verification system, could you explain me about why you add it and how you implemented it? `email-authcode`

The verification was implemented because of heirarchical datarooms, the things is for heirarchical datarooms a separate URL needed to be generated (for navigation purposes), the thing is since it was a separate URL it would bypass current verification system, that's was the reason for creation, but I will remove it since we don't need that.

**Comment by Aashish-Upadhyay-101** [comment]
> > I have almost reviewed your code, I just have one thing to ask, you added a new verification system, could you explain me about why you add it and how you implemented it? `email-authcode`
> 
> The verification was implemented because of heirarchical datarooms, the things is: for heirarchical datarooms a separate URL was needed to be generated (for navigation purposes), since it was a separate URL it would bypass current verification system, that's was the reason for creation, but I will remove it since we don't need that.

yes please do it and resolve the conflicts !

**Comment by mfts** [comment]
> However, regarding email-authcode I wanted ask @mfts that since email-verification to view document feature is for pro-users. Do we want to that for datarooms as well?
> If yes, can you please finalize that email-verification PR https://github.com/mfts/papermark/pull/197.

The email verification is already in production. 
You can copy/reuse this function: https://github.com/mfts/papermark/blob/77699f3404c538a25e4450f9d94bc933378cbb9c/pages/api/links/%5Bid%5D/verify-email.ts

**Comment by digant2482** [comment]
> > However, regarding email-authcode I wanted ask @mfts that since email-verification to view document feature is for pro-users. Do we want to that for datarooms as well?
> > If yes, can you please finalize that email-verification PR #197.
> 
> The email verification is already in production. You can copy/reuse this function: https://github.com/mfts/papermark/blob/77699f3404c538a25e4450f9d94bc933378cbb9c/pages/api/links/%5Bid%5D/verify-email.ts

@Aashish-Upadhyay-101, @mfts All set ready for review

**Pull request closed without merging**